Daikin decode support and enhancements

sillyfrog opened the pull request:
@crankyoldgit this includes updates to the changes you suggested.

I am not yet using the new AC extensions, but can work on that once this is merged in as I do not want to get further behind master.

sillyfrog committed:
Progress commit, working but with debug and incomplete

sillyfrog committed:
Progress commit, working but with debug and incomplete

sillyfrog committed:
Timers not working

sillyfrog committed:
Timers not working

sillyfrog committed:
Working Daikin decode, updated send & timer support

sillyfrog committed:
Working Daikin decode, updated send & timer support

sillyfrog committed:
Remove  set overflow

sillyfrog committed:
Remove  set overflow

crankyoldgit committed:
Dev branch for the 2.3.0 release.

crankyoldgit committed:
Update README information
* For v2.3
* General cleanup.
* More instructions and links.

crankyoldgit committed:
Add support for decoding Kelvinator A/C messages. (#332)

* Add support for decoding Kelvinator A/C messages.
* Add a decodeKelvinator() method.
* Tweak Kelvinator timings to use a large-ish common tick size.
* Unit test(s) for decodeKelvinator().
* Use a union for state to use less memory.
* Add support for Kelvinator A/C decoding to the dump example code.

crankyoldgit committed:
Initial support for sending Toshiba A/C messages. (#333)

* Initial support for sending Toshiba A/C messages.
* Add known-working A/C unit model info.

crankyoldgit committed:
Add support for all A/C units in MQTT example code.

crankyoldgit committed:
Update the AirCon HTML form to send all complex A/C types.

crankyoldgit committed:
Correct missing Fujitsu A/C functionality.

sillyfrog committed:
Cleanups and move header1 to flash

sillyfrog committed:
Cleanups and move header1 to flash

sillyfrog committed:
Merge branch 'mqtt_aircons' of github.com:markszabo/IRremoteESP8266

sillyfrog committed:
Don't include Arduion.h in tests

sillyfrog committed:
Don't include Arduion.h in tests

sillyfrog committed:
Use new State data when decoding

sillyfrog committed:
Use new State data when decoding

sillyfrog committed:
Merge branch 'master' of github.com:sillyfrog/IRremoteESP8266

crankyoldgit commented:
Yeah. This is going to take some work to merge in. Travis has a load of errors/issues.

I'll try to make some direct changes in your repo/PR rather than comments here. After I've done that, I'll go through and review the code.

Agreed, lets merge this into 'master' and then send you a PR to review for cleaning it up to the 'v2.3.0-dev' branch.

crankyoldgit commented:
Oh, I see you are still sending in patches. I'll wait a bit before I start.

crankyoldgit commented:
@sillyfrog On second thoughts, can you do your PR against the v2.3.0-dev branch?
i.e. The PR as it currently exists is sucking in a lot of the v2.3.0-dev changes into 'master'.

sillyfrog commented:
I’ll be back in a few hours and will give that a go. GitHub PRs like this do my head in :)

I have stopped pushing for the moment, but it now includes your new state field changes (I didn’t mean to add to the same PR, but I think it’s actually better)

Cheers

> On 12 Nov 2017, at 1:58 pm, David Conran <notifications@github.com> wrote:
> 
> @sillyfrog On second thoughts, can you do your PR against the v2.3.0-dev branch?
> i.e. The PR as it currently exists is sucking in a lot of the v2.3.0-dev changes into 'master'.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

crankyoldgit commented:
I fully understand what you mean w.r.t. these kinds of PRs. It seems to snowball out of control.
I've often resorted to trying merge, then reverting back to where I started, `rebase`ing and trying to roll my changes forward to get it to merge cleanly. Good luck. Sing out if you need a hand/advice/shoulder to cry on. :-)

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -3,13 +3,14 @@ An Arduino sketch to emulate IR Daikin ARC433** remote control unit
 http://harizanov.com/2012/02/control-daikin-air-conditioner-over-the-internet/
 
-Copyright 2016 sillyfrog
+Copyright 2017 sillyfrog
```
Leave the earlier year. IIRC earlier time/year is better. Feel free to change it to "2016, 2017" if you want .. and it doesn't make the cpplint.py program complain. ;-)

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -46,20 +67,20 @@ void IRsend::sendDaikin(unsigned char data[], uint16_t nbytes,
+             DAIKIN_ZERO_SPACE, 0, 5, false);
+    sendDaikinGapHeader();
+    // Leading header
+    // Do this as a constant to save RAM and keep in flash memory
```
Nice!!

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -191,25 +193,28 @@ uint8_t IRDaikinESP::getTemp() {
-  fan -= 2;
-  if (fan > DAIKIN_FAN_MAX)
-    fan = DAIKIN_FAN_AUTO;
+  if (fan == DAIKIN_FAN_QUITE || fan == DAIKIN_FAN_AUTO)
```
It's poor form to have a "pass" block.
You could change the logic to:
```
  if (fan != DAIKIN_FAN_QUITE && fan != DAIKIN_FAN_AUTO)
    fan -= 2;
```

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -234,7 +239,9 @@ void IRDaikinESP::setMode(uint8_t mode) {
       mode = DAIKIN_AUTO;
   }
-  daikin[13] = (mode << 4) | getPower();
+  mode = mode << 4;
```
Style suggestion only: `mode <<= 4;`

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+
+
+void IRDaikinESP::clearBit(uint8_t byte, uint8_t bitmask) {
+  bitmask = ~bitmask;
```
You can simplify this to just:
`daikin[byte] &= ~bitmask;`

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+void IRDaikinESP::clearBit(uint8_t byte, uint8_t bitmask) {
+  bitmask = ~bitmask;
+  daikin[byte] &= bitmask;
+  checksum();
```
If we do `checksum()`s prior exporting the data (via returning a ptr, or sending the message) you don't need to do intermediate checksums, thus saving processing time.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  // only keep 4 bits
+  daikin[19] &= 0xF0;
+  daikin[19] |= starttime;
+  checksum();
```
See other comment on `checksum()`s

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+void IRDaikinESP::disableOnTimer() {
+  enableOnTimer(0x600);
+  daikin[13] &= 0b11111101;
+  checksum();
```
ditto

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  uint16_t lopbits;
+  lopbits = endtime;
+  lopbits &= 0x0FF0;
+  lopbits = lopbits >> 4;
```
lopbits >>= 4;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  lopbits = lopbits >> 4;
+  daikin[20] = lopbits;
+  endtime &= 0x000F;
+  endtime = endtime << 4;
```
endtime <<= 4;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  endtime = endtime << 4;
+  daikin[19] &= 0x0F;
+  daikin[19] |= endtime;
+  checksum();
```
See other comments.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+void IRDaikinESP::disableOffTimer() {
+  enableOffTimer(0x600);
+  daikin[13] &= 0b11111011;
+  checksum();
```
See other comments.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+
+void IRDaikinESP::disableOffTimer() {
+  enableOffTimer(0x600);
+  daikin[13] &= 0b11111011;
```
`clearBit(13, 4);` ?

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+uint16_t IRDaikinESP::getOffTime() {
+  uint16_t ret, tmp;
+  ret = daikin[20];
+  ret = ret << 4;
```
ret <<= 4;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  ret = daikin[20];
+  ret = ret << 4;
+  tmp = daikin[19] & 0xF0;
+  tmp = tmp >> 4;
```
tmp >>= 4;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+}
+
+bool IRDaikinESP::getOffTimerEnabled() {
+  return daikin[13] & 0b00000100;
```
`return getBit(13, 4);` ?

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  // only keep 4 bits
+  daikin[6] &= 0xF0;
+  daikin[6] |= time;
+  checksum();
```
Ditto other checksum comments.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+uint16_t IRDaikinESP::getCurrentTime() {
+  uint16_t ret;
+  ret = daikin[6] & 0x0F;
+  ret = ret << 8;
```
ret <<= 8;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  return ret;
+}
+
+String IRDaikinESP::renderTime(uint16_t timemins) {
```
IIRC, you need to make this Arduino-only code.
i.e. `String` is Arduino specific. It will break the unit tests. i.e. won't compile.
So wrap this method with either:
```
#ifndef UNIT_TEST
...
#endif  // UNIT_TEST
```
or
```
#ifdef ARDUINO
...
#endif  // ARDUINO
```

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+
+#if DAIKIN_DEBUG
+
+void IRDaikinESP::printState() {
```
Ditto the comment on the previous method.
`Serial` & `String` are Arduino-only/specific.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  uint32_t ret = 0;
+  uint32_t tmp = 0;
+  if (getPower())
+    ret |= 0b00000000000000000000000000000001;
```
You could just say:
`ret |= 1;`

:-P

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  if (getPower())
+    ret |= 0b00000000000000000000000000000001;
+  tmp = getMode();
+  tmp = tmp << 1;
```
tmp <<= 1;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  ret |= tmp;
+
+  tmp = getFan();
+  tmp = tmp << 4;
```
tmp <<= 4;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  ret |= tmp;
+
+  tmp = getTemp();
+  tmp = tmp << 8;
```
tmp <<= 8;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+  if (getSwingVertical())
+    ret |= 0b00000000000010000000000000000000;
+  uint32_t time = getCurrentTime();
+  time = time << 20;
```
time <<= 20;

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+    ret |= 0b00000000000010000000000000000000;
+  uint32_t time = getCurrentTime();
+  time = time << 20;
+  ret |= time;
```
or just `ret |= (getCurrentTime() << 20);` and save a variable.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,25 +268,457 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+// Decode the supplied Daikin A/C message.
 // Args:
 //   results: Ptr to the data to decode and where to store the decode result.
 //   nbits:   Nr. of bits to expect in the data portion. Typically SAMSUNG_BITS.
```
Oops. Looks like I left SAMSUNG_BITS. Can you please change it to DAIKIN_BITS? Thanks.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -288,60 +727,64 @@ bool IRrecv::decodeDaikin(decode_results *results, uint16_t nbits,
-
-  // Header
-  if (!matchMark(results->rawbuf[offset++], DAIKIN_HDR_MARK))
+  unsigned char daikin_code[DAIKIN_COMMAND_LENGTH + 2];
```
Question: Why is this + 2 longer than DAIKIN_COMMAND_LENGTH?
Should we make DAIKIN_COMMAND_LENGTH the correct value?

sillyfrog replied:
This is a (evil) hack so I can pass state data around by reference while decoding (see `DAIKIN_CURBIT` and `DAIKIN_CURINDEX`). So it's not actually required else where.

crankyoldgit replied:
Oh, you are right. That is evil. Too evil in fact, I think. I'll make some suggestions how to do this using readData(). That said, I like the bounds checking, which readData() doesn't have. Hmmm

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -288,60 +727,64 @@ bool IRrecv::decodeDaikin(decode_results *results, uint16_t nbits,
-    DPRINTLN("no gap");
-    return false;
-  }
+  yield();
```
I'm assuming this is here because it was watchdog resetting due to it taking so long to print all the debug info.
If so, we should probably only include it if DEBUG is set.
e.g.
```
#if DEBUG
  yield();
#endif
```

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -288,60 +727,64 @@ bool IRrecv::decodeDaikin(decode_results *results, uint16_t nbits,
+#endif
+
+  // Copy across the bits to state
+  for (uint8_t i = 0; i < DAIKIN_COMMAND_LENGTH; i++)
```
Still curious about the `+ 2` extra bytes? Are they the checksum or something?
If so, we should probably increase DAIKIN_COMMAND_LENGTH appropriately.

crankyoldgit commented on src/ir_Daikin.h:
```diff
@@ -4,6 +4,12 @@
+#endif
+
+// Option to disable the additional Daikin debug info to conserve memory
+#define DAIKIN_DEBUG 1
```
Probably should have this _off_ by default.

crankyoldgit reviewed (commented):
Sorry about the large amount of comments. Most of them are repetitive if that's any consolation.

crankyoldgit committed:
Clean up a bunch of cpplint.py issues.

crankyoldgit committed:
More linter/style issues fixed.

crankyoldgit commented:
@sillyfrog I've taken the liberty to commit a few cosmetic commits to fix up some style/issues the linter found.
I've also rebased your master branch (https://github.com/sillyfrog/IRremoteESP8266) against the v2.3.0-dev branch.
That's cleaned up a bunch of the commits in this PR.
However, you should probably abandon this PR, and create a new one against the 'v2.3.0-dev' as this one is targeted at 'master'.

crankyoldgit committed:
Fix a number of compiler warnings and allow unit tests to compile.
- Unit tests don't pass yet.

crankyoldgit commented on src/ir_Daikin.cpp:
```diff
@@ -261,87 +268,517 @@ void IRDaikinESP::setSwingHorizontal(bool state) {
+}
+
+uint16_t checkheader(decode_results *results, uint16_t offset,
+                     unsigned char daikin_code[]) {
```
`daikin_code[]` appears to be unused.

sillyfrog committed:
Minor enhancements and working example

sillyfrog committed:
Merge branch 'master' of github.com:sillyfrog/IRremoteESP8266

sillyfrog commented:
I have just pushed a number of changes (and merged the conflicts), but I'm out for tonight. I'll try and get back to things in the next few days. I have no idea where this is going to merge now or how it will look :/
Hopefully it's not too much of a mess. Let me know if you see any issues!
Cheers.

crankyoldgit commented:
@sillyfrog It's getting there. Don't worry.

I'm happy with the decode stuff being broken for daikin, but I'm slightly concerned that the output of sendDaikin() is now vastly different to previous, and to what I'd expect of a typical IR protocol.
e.g. most protocols start with a large (long) header (mark + space). It now seems to start with a sequence of 'data' bits, then it progresses into a normal-ish header.

Any chance you can post a complete raw dump of a typical IR message from the remote? Bonus points for detailing what the settings are on the remote etc.

crankyoldgit committed:
Move a bunch of #define's to the header file.
Fix a few linter/style issues as reported by cpplint.py

sillyfrog commented:
Yeah, I'm amazed it worked given how much was missing.

Here is the output from a capture when I was testing last night (with DAIKIN_DEBUG turned on, which I committed, which I didn't mean to):
```Raw Bits:
00010001 11011010 00100111 00000000 01000010 00111010 00000101 10010011 00010001 11011010 00100111 00000000 00000000 00111111 00111010 00000000 10100000 00000000 00001010 00100101 00010111 00000001 00000000 11000000 00000000 00000000 00110010 
Power: On
Temperature: 29
Quiet: Off
Sensor: Off
Powerful: On
Econo: Off
Eye: Off
Mold: Off
Swing Vertical: Off
Swing Horizontal: Off
Fan Speed: Auto
Mode: Cool
On Timer: Enabled
On Time: 21:30
Off Timer: Enabled
Off Time: 6:10
Current Time: 22:18
Encoding  : DAIKIN
Code      : 11DA2700423A059311DA2700003F3A00A0000A25170100C0000032 (216 bits)
Timing[583]: 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   418, -   446, 
   +   416, -   446,    +   416, - 25434,    +  3436, -  1768,    +   390, -  1336, 
   +   390, -   446,    +   416, -   446,    +   416, -   446,    +   416, -  1336, 
   +   390, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -  1336,    +   390, -   448,    +   416, -  1336,    +   390, -  1336, 
   +   390, -   448,    +   416, -  1336,    +   390, -  1336,    +   390, -  1338, 
   +   388, -  1338,    +   390, -  1336,    +   390, -   446,    +   416, -   446, 
   +   416, -  1336,    +   390, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   448, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -  1336, 
   +   390, -   446,    +   416, -  1336,    +   390, -   448,    +   416, -   446, 
   +   416, -   446,    +   416, -  1336,    +   390, -  1336,    +   390, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   448,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   448,    +   414, -   448,    +   416, -   448,    +   416, -  1336, 
   +   390, -  1336,    +   390, -  1336,    +   390, -   446,    +   414, -  1336, 
   +   390, -   448,    +   414, -  1336,    +   390, -  1336,    +   390, - 34878, 
   +  3436, -  1768,    +   390, -  1336,    +   390, -   446,    +   416, -   448, 
   +   416, -   446,    +   416, -  1336,    +   390, -   446,    +   416, -   448, 
   +   416, -   446,    +   416, -   446,    +   416, -  1336,    +   390, -   446, 
   +   416, -  1336,    +   390, -  1336,    +   390, -   446,    +   416, -  1336, 
   +   390, -  1336,    +   390, -  1336,    +   390, -  1336,    +   390, -  1336, 
   +   392, -   446,    +   414, -   448,    +   416, -  1336,    +   390, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   414, -   448, 
   +   416, -   446,    +   416, -   448,    +   414, -   448,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   414, -  1336,    +   390, -   448, 
   +   416, -   446,    +   416, -   446,    +   416, -   448,    +   416, -  1336, 
   +   390, -   446,    +   416, -   446,    +   416, -  1336,    +   390, -   446, 
   +   416, -  1336,    +   390, -  1336,    +   390, -  1336,    +   390, -   446, 
   +   416, -   446,    +   414, -  1338,    +   390, -   446,    +   416, -  1336, 
   +   390, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -  1336,    +   390, -  1336,    +   390, -   446, 
   +   416, -   446,    +   416, -  1336,    +   390, -   446,    +   416, -   446, 
   +   416, -  1336,    +   390, - 34876,    +  3436, -  1768,    +   388, -  1336, 
   +   390, -   446,    +   416, -   446,    +   416, -   448,    +   416, -  1336, 
   +   390, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   448, 
   +   416, -  1336,    +   390, -   448,    +   414, -  1336,    +   390, -  1336, 
   +   390, -   446,    +   416, -  1336,    +   388, -  1338,    +   388, -  1336, 
   +   390, -  1336,    +   390, -  1336,    +   390, -   446,    +   416, -   446, 
   +   416, -  1336,    +   390, -   446,    +   420, -   442,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   448,    +   416, -   446,    +   416, -   448, 
   +   416, -   446,    +   416, -   448,    +   416, -   446,    +   416, -  1336, 
   +   390, -  1336,    +   390, -  1336,    +   388, -  1338,    +   390, -  1336, 
   +   390, -  1336,    +   392, -   446,    +   416, -   446,    +   416, -   448, 
   +   416, -  1334,    +   390, -   446,    +   416, -  1338,    +   388, -  1336, 
   +   390, -  1336,    +   390, -   446,    +   416, -   446,    +   416, -   448, 
   +   414, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   448, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -  1336,    +   390, -   446,    +   416, -  1336,    +   390, -   446, 
   +   414, -   448,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   448,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -  1336,    +   390, -   446,    +   416, -  1336,    +   390, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   448,    +   416, -  1338, 
   +   390, -   444,    +   418, -  1336,    +   390, -   448,    +   416, -   446, 
   +   416, -  1336,    +   390, -   446,    +   416, -   446,    +   416, -  1336, 
   +   390, -  1336,    +   388, -  1336,    +   390, -   446,    +   416, -  1336, 
   +   390, -   448,    +   414, -   448,    +   414, -   448,    +   416, -  1334, 
   +   390, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   448, 
   +   416, -   446,    +   416, -   446,    +   416, -   448,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   416, -   446, 
   +   416, -   448,    +   416, -  1336,    +   390, -  1336,    +   390, -   446, 
   +   416, -   446,    +   416, -   446,    +   416, -   446,    +   414, -   446, 
   +   416, -   448,    +   416, -   446,    +   416, -   448,    +   414, -   446, 
   +   418, -   446,    +   416, -   446,    +   416, -   448,    +   416, -   446, 
   +   416, -   448,    +   416, -   446,    +   416, -   448,    +   416, -   446, 
   +   416, -  1336,    +   390, -   446,    +   416, -   446,    +   416, -  1338, 
   +   390, -  1336,    +   390, -   446,    +   416, -   446,    +   416
uint16_t rawData[583] = {416, 446,  416, 446,  416, 446,  418, 446,  416, 446,  416, 25434,  3436, 1768,  390, 1336,  390, 446,  416, 446,  416, 446,  416, 1336,  390, 446,  416, 446,  416, 446,  416, 446,  416, 1336,  390, 448,  416, 1336,  390, 1336,  390, 448,  416, 1336,  390, 1336,  390, 1338,  388, 1338,  390, 1336,  390, 446,  416, 446,  416, 1336,  390, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 448,  416, 446,  416, 446,  416, 446,  416, 1336,  390, 446,  416, 1336,  390, 448,  416, 446,  416, 446,  416, 1336,  390, 1336,  390, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 448,  416, 446,  416, 446,  416, 446,  416, 448,  414, 448,  416, 448,  416, 1336,  390, 1336,  390, 1336,  390, 446,  414, 1336,  390, 448,  414, 1336,  390, 1336,  390, 34878,  3436, 1768,  390, 1336,  390, 446,  416, 448,  416, 446,  416, 1336,  390, 446,  416, 448,  416, 446,  416, 446,  416, 1336,  390, 446,  416, 1336,  390, 1336,  390, 446,  416, 1336,  390, 1336,  390, 1336,  390, 1336,  390, 1336,  392, 446,  414, 448,  416, 1336,  390, 446,  416, 446,  416, 446,  416, 446,  414, 448,  416, 446,  416, 448,  414, 448,  416, 446,  416, 446,  416, 446,  414, 1336,  390, 448,  416, 446,  416, 446,  416, 448,  416, 1336,  390, 446,  416, 446,  416, 1336,  390, 446,  416, 1336,  390, 1336,  390, 1336,  390, 446,  416, 446,  414, 1338,  390, 446,  416, 1336,  390, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 1336,  390, 1336,  390, 446,  416, 446,  416, 1336,  390, 446,  416, 446,  416, 1336,  390, 34876,  3436, 1768,  388, 1336,  390, 446,  416, 446,  416, 448,  416, 1336,  390, 446,  416, 446,  416, 446,  416, 448,  416, 1336,  390, 448,  414, 1336,  390, 1336,  390, 446,  416, 1336,  388, 1338,  388, 1336,  390, 1336,  390, 1336,  390, 446,  416, 446,  416, 1336,  390, 446,  420, 442,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 448,  416, 446,  416, 448,  416, 446,  416, 448,  416, 446,  416, 1336,  390, 1336,  390, 1336,  388, 1338,  390, 1336,  390, 1336,  392, 446,  416, 446,  416, 448,  416, 1334,  390, 446,  416, 1338,  388, 1336,  390, 1336,  390, 446,  416, 446,  416, 448,  414, 446,  416, 446,  416, 446,  416, 448,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 1336,  390, 446,  416, 1336,  390, 446,  414, 448,  416, 446,  416, 446,  416, 446,  416, 448,  416, 446,  416, 446,  416, 446,  416, 1336,  390, 446,  416, 1336,  390, 446,  416, 446,  416, 446,  416, 448,  416, 1338,  390, 444,  418, 1336,  390, 448,  416, 446,  416, 1336,  390, 446,  416, 446,  416, 1336,  390, 1336,  388, 1336,  390, 446,  416, 1336,  390, 448,  414, 448,  414, 448,  416, 1334,  390, 446,  416, 446,  416, 446,  416, 448,  416, 446,  416, 446,  416, 448,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 446,  416, 448,  416, 1336,  390, 1336,  390, 446,  416, 446,  416, 446,  416, 446,  414, 446,  416, 448,  416, 446,  416, 448,  414, 446,  418, 446,  416, 446,  416, 448,  416, 446,  416, 448,  416, 446,  416, 448,  416, 446,  416, 1336,  390, 446,  416, 446,  416, 1338,  390, 1336,  390, 446,  416, 446,  416};  // DAIKIN 93053A420027DA11
uint8_t state[27] = {0x11, 0xDA, 0x27, 0x00, 0x42, 0x3A, 0x05, 0x93, 0x11, 0xDA, 0x27, 0x00, 0x00, 0x3F, 0x3A, 0x00, 0xA0, 0x00, 0x0A, 0x25, 0x17, 0x01, 0x00, 0xC0, 0x00, 0x00, 0x32};

```

crankyoldgit committed:
Rework checkheader() & readbits(). Add unit tests for decoding.
* Refactor checkheader() & readbits() to use a pointer for offset.
  This cleans up the code quite a lot & makes it more readable IMHO.
* Add some unit tests for decodeDaikin() using real and synthetic data.
  This allowed the refactoring above to happen without causing issues.
  Note: Not all (old) unit tests are passing at this point.
* Clean up some #define usage to make it more presentable and useful.
* General code clean up (lint and style etc)

crankyoldgit committed:
Fix setting econo, quiet, & powerful modes.

Unit tests were failing because econo, quiet, & powerful modes were
not mutually exclusive. Previous code did this.
The manual confirms this is correct. i.e. setting one of these modes cancels
the other conflicting modes.

The relevant unit tests now pass again.

crankyoldgit committed:
Update unit tests for getFan() & setFan().
- Update test cases to reflect new usage etc.
- Minor code cleanup.

crankyoldgit committed:
Powerful & Quiet, and Powerful & Econo are incompatible. Not Quiet & Econo.
@sillyfrog correctly pointed out Econo mode & Quiet mode are not mutually
exclusive.
Created unit tests for those cases, as well as created unit tests for
Econo mode.

crankyoldgit committed:
Update old unit test output to new daikin message protocol.
* Replace old formatted unit test data with data based on current sendDaikin()
* The unit test routines need more capacity to handle larger outputs.
  e.g. sendDaikin() with repeats has more than 1000 entries. Bumped to 10k.

crankyoldgit committed:
Fix Time/Timer methods.
* Assigning 16bit values on to a char array caused issues on some architectures.
* Cleaned up time(r) methods and correctly cleared/masked bits.
* Unit Tests added for the time/timer related methods.
* Turned off the DAIKIN_DEBUG flag.
* Minor code quality cleanups.

crankyoldgit committed:
Zero entire state on reset.
- We were missing zero'ing of daikin[3].

crankyoldgit committed:
Add unit test coverage for Sensor, Mold, & Eye modes.

crankyoldgit committed:
Set TIMEOUT to something more appropriate for A/C decoding.
This reverts/alters an earlier change.

crankyoldgit commented:
@sillyfrog this PR now cleanly merges with v2.3.0-dev, and passes all the unit tests & travis checks.
I'm happy to merge it now. However, before I do so, can you please cast an eye over it as it stands and check & confirm it works with your Daikin unit etc etc.

sillyfrog commented:
I have done some basic testing (both send and receive), and it's looking good! So merging in now.

Thanks for your help!
